File: Java/dynamic_programming/EditDistance.java
/**
 * Edit Distance Algorithm
 *
 * Description: Finds the minimum number of operations (insert, delete, or substitute)
 * required to convert one string into another using Dynamic Programming.
 * Time Complexity: O(m*n), where m and n are the lengths of the two strings.
 * Space Complexity: O(m*n) for the DP table.
 */
public class EditDistance {

    /**
     * Calculates the minimum edit distance between two words.
     *
     * @param word1 The first word.
     * @param word2 The second word.
     * @return The minimum number of operations to transform word1 to word2.
     */
    public static int minDistance(String word1, String word2) {
        int m = word1.length();
        int n = word2.length();

        // Create a DP table to store results of subproblems.
        // dp[i][j] will be the edit distance of the first i chars of word1 and first j chars of word2.
        int[][] dp = new int[m + 1][n + 1];

        // Fill the DP table in a bottom-up manner.
        for (int i = 0; i <= m; i++) {
            for (int j = 0; j <= n; j++) {
                // If the first string is empty, the only option is to insert all characters of the second string.
                if (i == 0) {
                    dp[i][j] = j;
                }
                // If the second string is empty, the only option is to remove all characters of the first string.
                else if (j == 0) {
                    dp[i][j] = i;
                }
                // If the last characters are the same, no new operation is needed.
                // The cost is the same as the cost for the strings without the last character.
                else if (word1.charAt(i - 1) == word2.charAt(j - 1)) {
                    dp[i][j] = dp[i - 1][j - 1];
                }
                // If the last characters are different, we must perform an operation.
                // We take the minimum cost of the three possible operations:
                // 1. Insert: dp[i][j-1]
                // 2. Delete: dp[i-1][j]
                // 3. Substitute: dp[i-1][j-1]
                else {
                    dp[i][j] = 1 + Math.min(dp[i][j - 1],      // Insert
                                           Math.min(dp[i - 1][j],      // Remove
                                                    dp[i - 1][j - 1])); // Replace
                }
            }
        }

        return dp[m][n];
    }

    // Test cases
    public static void main(String[] args) {
        // Test case 1: LeetCode example
        String word1 = "horse";
        String word2 = "ros";
        System.out.println("Edit distance between '" + word1 + "' and '" + word2 + "': " + minDistance(word1, word2)); // Expected Output: 3

        // Test case 2: Another LeetCode example
        String word3 = "intention";
        String word4 = "execution";
        System.out.println("Edit distance between '" + word3 + "' and '" + word4 + "': " + minDistance(word3, word4)); // Expected Output: 5

        // Test case 3: One string is empty
        String word5 = "apple";
        String word6 = "";
        System.out.println("Edit distance between '" + word5 + "' and '" + word6 + "': " + minDistance(word5, word6)); // Expected Output: 5

        // Test case 4: Both strings are identical
        String word7 = "test";
        String word8 = "test";
        System.out.println("Edit distance between '" + word7 + "' and '" + word8 + "': " + minDistance(word7, word8)); // Expected Output: 0
    }
}

File: Python/dynamic_programming/Edit_Distance.py
# Given two strings word1 and word2, return the minimum number of operations required to convert word1 to word2.
# You have the following three operations permitted on the string word1:

# Insert a character
# Delete a character
# Replace a character


# ask:
# Since this will be a functional problem, you don't have to take input. You just have to complete the function minDistance() that takes two strings word1 and word2 as its parameters.

# Custom Input:
# The input consists of two lines:
# The first line contains the initial string word1.
# The second line contains the target string word2.



def minDistance(word1, word2):
    m, n = len(word1), len(word2)
    dp = [[0]*(n+1) for i in range(m+1)]
    for i in range(m+1):
        dp[i][0] = i
    for j in range(n+1):
        dp[0][j] = j
    for i in range(1, m+1):
        for j in range(1, n+1):
            if word1[i-1] == word2[j-1]:
                dp[i][j] = dp[i-1][j-1]
            else:
                dp[i][j] = 1 + min(dp[i-1][j], dp[i][j-1], dp[i-1][j-1])
    return dp[m][n]








# Input:
family
smile
# Output:
3
Explanation:
family → amily (delete 'f')
amily → smily (replace 'a' with 's')
smily → smile (replace 'y' with 'e')

# Input:
intention
execution
# Output:
5
# Explanation
intention → inention (remove 't')
inention → enention (replace 'i' with 'e')
enention → exention (replace 'n' with 'x')
exention → exection (replace 'n' with 'c')
exection → execution (insert 'u')